atlasdb-config/src/main/java/com/palantir/atlasdb/config/DiscoverableSubtypeResolver.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright 2018 Palantir Technologies, Inc. All rights reserved.
+ *
+ * Licensed under the BSD-3 License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://opensource.org/licenses/BSD-3-Clause
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
```

Comment:
So this is copied from Dropwizard's [DiscoverableSubtypeResolver](https://github.com/dropwizard/dropwizard/blob/master/dropwizard-jackson/src/main/java/io/dropwizard/jackson/DiscoverableSubtypeResolver.java) - which means we'll need to get the licensing correct. In particular this means we'll need to include a copy of the Apache licence - see [RateLimiter](https://github.com/palantir/atlasdb/blob/c363b57274799c5430ac7cff38b67b94dfc8d188/qos-service-impl/src/main/java/com/palantir/atlasdb/qos/ratelimit/guava/RateLimiter.java) for something similar we did. Happy to help with implementing this if needed!